libcentrifugo/conns/conns.go

```diff
@@ -3,8 +3,50 @@ package conns
 import (
 	"encoding/json"
 	"sync"
+
+	"github.com/gorilla/websocket"
 )
 
+// QueuedMessage is a wrapper structure over raw data payload.
+// It can optionally contain prepared websocket message (to drastically
+// reduce overhead of framing in case of using websocket compression).
+type QueuedMessage struct {
+	// Payload is raw message payload (encoded in JSON).
+	Payload []byte
+	// UsePrepared can be used in session Send method to determine which
+	// underlying transport write method to use. At moment used for websocket
+	// session only - so we can call WriteMessage or WritePreparedMessage
+	// connection methods.
+	UsePrepared bool
+	prepared    *websocket.PreparedMessage
+	once        sync.Once
+}
+
+// NewQueuedMessage initializes QueuedMessage.
+func NewQueuedMessage(payload []byte, usePrepared bool) *QueuedMessage {
+	m := &QueuedMessage{
+		Payload:     payload,
+		UsePrepared: usePrepared,
+	}
+	return m
+}
+
+// Prepared allows to get PreparedMessage for raw websocket connections. It
+// constructs PreparedMessage lazily after first call.
+func (m *QueuedMessage) Prepared() *websocket.PreparedMessage {
+	m.once.Do(func() {
+		pm, _ := websocket.NewPreparedMessage(websocket.TextMessage, m.Payload)
+		m.prepared = pm
+	})
+	return m.prepared
+}
+
+// Len returns length of QueuedMessage payload so QueuedMessage implements
+// item that can be queued into our unbounded queue.
```

Comment:
The queue implementation itself is unbounded, we check it's length in caller's code before adding message.